query_optimizer/logical/TableGenerator.hpp

```diff
@@ -0,0 +1,176 @@
+/**
+ *   Copyright 2016 Pivotal Software, Inc.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_QUERY_OPTIMIZER_LOGICAL_TABLE_GENERATOR_HPP_
+#define QUICKSTEP_QUERY_OPTIMIZER_LOGICAL_TABLE_GENERATOR_HPP_
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "expressions/table_generator/GeneratorFunctionHandle.hpp"
+#include "query_optimizer/OptimizerTree.hpp"
+#include "query_optimizer/expressions/AttributeReference.hpp"
+#include "query_optimizer/logical/Logical.hpp"
+#include "query_optimizer/logical/LogicalType.hpp"
+#include "utility/Cast.hpp"
+#include "utility/Macros.hpp"
+
+#include "glog/logging.h"
+
+namespace quickstep {
+namespace optimizer {
+namespace logical {
+
+namespace E = ::quickstep::optimizer::expressions;
+
+using quickstep::GeneratorFunctionHandle;
+
+/** \addtogroup OptimizerLogical
+ *  @{
+ */
+
+class TableGenerator;
+typedef std::shared_ptr<const TableGenerator> TableGeneratorPtr;
+
+/**
+ * @brief Leaf logical node that represents a table that will be populated
```

Comment:
I added a GENERATOR_FUNCTIONS.md file that describes the architecture of generator functions. Perhaps let's read that file first and decide if it is best to also put the descriptions here.